Simple decoupled file based policy engine

johnandersen777 opened the pull request:
Simple insert policy based engine based on presence of `operation.policy.{insert,denied,failed}` files. Currently only for use with `use_lro=True`. This is a simple way to enable evaluation of claims prior to submission by arbitrary policy engines which watch the workspace (fanotify,  inotify, etc.).

### [Jump to viewing docs](https://github.com/scitt-community/scitt-api-emulator/pull/27/files?short_path=902592d#diff-902592dae6c982dcbc114dc4619188a03c4f5c5f27246b60aca5a1e1c655e1b6)

- References
  - [Associated mailing list thread](https://mailarchive.ietf.org/arch/msg/scitt/PwcesOR1txbHOIjMhMOt01YWDjo/)
  - https://github.com/ietf-scitt/use-cases/pull/18
    - > Recent discussion has revealed most of this use case will be focused on the policy / gatekeeper component and federation components of SCITT.
- TODO
  - [x] Unittests
  - [x] Docs
    - [x] [asciicast](https://github.com/asciinema)
  - [x] failure -> failed typos
  - [x] error.detail
    - Would be great if `detail` was optionally an object. A string is of limited usefulness, the same issue was had with the SPDX 2.X series with regards to description fields.
      - https://github.com/ietf-wg-scitt/draft-ietf-scitt-architecture/pull/60#discussion_r1148486103
      - https://github.com/ietf-wg-scitt/draft-ietf-scitt-architecture/issues/62
    - https://mailarchive.ietf.org/arch/msg/scitt/c0t5zLUJtCQ9_Jrf7mykWXSIn94/
      - Future
        - Updating `operationId` to `operationUrl` is out of scope of this pull request.
- Future
  - Workflow based policy execution within documentation rather than simple bash example

pdxjohnny committed:
Simple file presence based policy evaluation

Signed-off-by: John Andersen <johnandersenpdx@gmail.com>

pdxjohnny committed:
Document simple decoupled file based policy engine

Signed-off-by: John Andersen <johnandersenpdx@gmail.com>

pdxjohnny committed:
Tests for simple decoupled file based policy engine

Signed-off-by: John Andersen <johnandersenpdx@gmail.com>

pdxjohnny committed:
asciicast example simple decoupled file based policy engine for documentation

Signed-off-by: John Andersen <johnandersenpdx@gmail.com>

pdxjohnny committed:
Implement reason communication associated with policy enforcement

- Updated simple file based policy engine to align with @darrelmiller review of SCITT architecture documentation
  - His full review can be found at https://mailarchive.ietf.org/arch/msg/scitt/c0t5zLUJtCQ9_Jrf7mykWXSIn94/
- Do not attempt to load policy failed/denied JSON information if present triggering file empty

Signed-off-by: John Andersen <john.s.andersen@intel.com>

pdxjohnny commented on scitt_emulator/scitt.py:
```diff
@@ -168,9 +169,23 @@ def _sync_policy_result(self, operation: dict):
+                RecursionError,
+                json.JSONDecodeError,
+            ):
+                policy_result.update(json.loads(policy_denied_path.read_text()))
```
I don't currently see logging infrastructure but we should add some here if we add logging code.

johnandersen777 commented on docs/registration_policies.md:
````diff
@@ -0,0 +1,131 @@
+
+```console
+$ npm install -g nodemon
+$ echo '{"reason": "blocklisted issuer"}' | tee reason.json
````
Per https://mailarchive.ietf.org/arch/msg/scitt/c0t5zLUJtCQ9_Jrf7mykWXSIn94/

```suggestion
$ echo '{"type": "content_address_of_policy", "detail": "blocklisted issuer"}' | tee reason.json
```

johnandersen777 commented on docs/registration_policies.md:
```diff
@@ -0,0 +1,131 @@
+    raise_for_operation_status(operation)
+  File "/home/alice/Documents/python/scitt-api-emulator/scitt_emulator/client.py", line 29, in raise_for_operation_status
+    raise RuntimeError(f"Operation error: {operation['error']}")
+RuntimeError: Operation error: {reason': 'blocklisted issuer', 'status': 'denied'}
```
Per https://mailarchive.ietf.org/arch/msg/scitt/c0t5zLUJtCQ9_Jrf7mykWXSIn94/

```suggestion
RuntimeError: Operation error: {'error': {'type': 'content_address_of_policy', 'detail': 'blocklisted issuer'}, 'status': 'denied'}
```

pdxjohnny commented on scitt_emulator/scitt.py:
```diff
@@ -142,11 +157,52 @@ def _create_operation(self, claim: bytes):
+                raise PolicyResultDecodeError(operation_id) from error
+            policy_failed_path.unlink()
+        if policy_denied_path.exists():
+            policy_result["status"] = "denied"
```
Per https://mailarchive.ietf.org/arch/msg/scitt/c0t5zLUJtCQ9_Jrf7mykWXSIn94/

```suggestion
            policy_result["type"] = "denied"
```

pdxjohnny commented on scitt_emulator/scitt.py:
```diff
@@ -142,11 +157,52 @@ def _create_operation(self, claim: bytes):
+        if policy_denied_path.exists():
+            policy_result["status"] = "denied"
+            try:
+                policy_result.update(json.loads(policy_denied_path.read_text()))
+            except Exception as error:
+                raise PolicyResultDecodeError(operation_id) from error
```
```suggestion
                policy_result_error = json.loads(policy_denied_path.read_text())
            except Exception as error:
                raise PolicyResultDecodeError(operation_id) from error
            policy_result["error"] = policy_result_error
```

johnandersen777 commented on scitt_emulator/scitt.py:
```diff
@@ -142,11 +157,52 @@ def _create_operation(self, claim: bytes):
+        if policy_denied_path.exists():
+            policy_result["status"] = "denied"
+            try:
+                policy_result.update(json.loads(policy_denied_path.read_text()))
```
Per https://mailarchive.ietf.org/arch/msg/scitt/c0t5zLUJtCQ9_Jrf7mykWXSIn94/

```suggestion
                policy_result_error = json.loads(policy_denied_path.read_text())
```

johnandersen777 commented on scitt_emulator/scitt.py:
```diff
@@ -142,11 +157,52 @@ def _create_operation(self, claim: bytes):
+                policy_result.update(json.loads(policy_denied_path.read_text()))
+            except Exception as error:
+                raise PolicyResultDecodeError(operation_id) from error
+            policy_denied_path.unlink()
```
Per https://mailarchive.ietf.org/arch/msg/scitt/c0t5zLUJtCQ9_Jrf7mykWXSIn94/

```suggestion
            policy_result["error"] = policy_result_error
            policy_denied_path.unlink()
```

johnandersen777 commented on scitt_emulator/scitt.py:
```diff
@@ -142,11 +157,52 @@ def _create_operation(self, claim: bytes):
+        if policy_failed_path.exists():
+            policy_result["status"] = "failed"
+            try:
+                policy_result.update(json.loads(policy_failed_path.read_text()))
```
Per https://mailarchive.ietf.org/arch/msg/scitt/c0t5zLUJtCQ9_Jrf7mykWXSIn94/

```suggestion
                policy_result_error = json.loads(policy_denied_path.read_text())
```

johnandersen777 commented on scitt_emulator/scitt.py:
```diff
@@ -142,11 +157,52 @@ def _create_operation(self, claim: bytes):
+                policy_result.update(json.loads(policy_failed_path.read_text()))
+            except Exception as error:
+                raise PolicyResultDecodeError(operation_id) from error
+            policy_failed_path.unlink()
```
Per https://mailarchive.ietf.org/arch/msg/scitt/c0t5zLUJtCQ9_Jrf7mykWXSIn94/

```suggestion
            policy_result["error"] = policy_result_error
            policy_failed_path.unlink()
```

johnandersen777 commented on docs/registration_policies.md:
```diff
@@ -0,0 +1,131 @@
+## Simple decoupled file based policy engine
+
+The SCITT API emulator can deny entry based on presence of
+`operation.policy.{insert,denied,failure}` files. Currently only for use with
```
```suggestion
`operation.policy.{insert,denied,failed}` files. Currently only for use with
```

johnandersen777 commented on docs/registration_policies.md:
```diff
@@ -0,0 +1,131 @@
+    raise_for_operation_status(operation)
+  File "/home/alice/Documents/python/scitt-api-emulator/scitt_emulator/client.py", line 29, in raise_for_operation_status
+    raise RuntimeError(f"Operation error: {operation['error']}")
+RuntimeError: Operation error: {'error': {'detail': 'blocklisted issuer', 'type': 'content_address_of_policy'}, 'status': 'denied'}
```
Per https://mailarchive.ietf.org/arch/msg/scitt/c0t5zLUJtCQ9_Jrf7mykWXSIn94/

```suggestion
RuntimeError: Operation error: {'error': {'detail': 'blocklisted issuer', 'type': 'content_address_of_policy'}, 'status': 'failed'}
```

johnandersen777 commented:
This should be ready for review

johnandersen777 commented:
Pinging owners for review please @henkbirkholz @JAG-UK @kkarunakaran89 @letmaik @mprorock @OR13 @SteveLasker @yogeshbdeshpande, thank you

OR13 reviewed (dismissed):
Seems like this PR is an implementation of policy configuration based on python files.

It might be preferable to use a different policy language to communicate the decoupled nature in the future, but support for COSE could be a blocker for that.

I think we have considered specific policies or policy engines out of scope, but we do seem to need an ability to identify policy files in use.

Is there a way we can surface "URNs" for example policies you have provided?

johnandersen777 commented:
Thank you for the review!

In the example documentation the policy engine is indeed implemented with python files. However this implementation lives only within that documentation file (although it's currently used within the associated tests). The intent here was to show that the execution of the engine is arbitrary, as far as the SCITT API Emulator is concerned. Policy engine results are communicated through the the presence of files and the reasoning for denial or failure potentially through their contents (`reason.json` in the example docs and tests).

The plan is to expand on this pull request in a future pull request to leverage policy referenced as content addressable JSON based abstraction of the policy, for example: `did:web:registry.example.com:policy-as-code:blocklist%40sha256%3Aaaaaaaaa`. The policy could then itself be secured following the [Distributing with OCI Registries, with SCITT Enhancements](https://scitt.io/distributing-with-oci-scitt.html) document.

> One could also leverage domain specific languages rather than a generic JSON for policy definition (examples: [seedwing dogma](https://docs.seedwing.io/seedwing/policies/openvex/index.html#from-guac), [open policy agent](https://www.openpolicyagent.org/docs/latest/cli/)).

```console
$ export REGISTRY=registry.example.com
$ export PROJECT=policy-as-code
$ export IMAGE=blocklist
$ export DIGEST=sha256:aaaaaaaa
$ export DID_WEB_OF_POLICY=$(python -c 'import sys, urllib; print(urllib.parse.quote(sys.argv[-1]))' "${IMAGE}@${DIGEST}")
$ echo "$(cat workspace/service_parameters.json)" \
    | jq ".insertPolicy = \"did:web:${REGISTRY}:${PROJECT}:${DID_WEB_OF_POLICY}\"" \
    | tee workspace/service_parameters.json.new \
    && mv workspace/service_parameters.json.new workspace/service_parameters.json
{
  "serviceId": "emulator",
  "treeAlgorithm": "CCF",
  "signatureAlgorithm": "ES256",
  "serviceCertificate": "-----BEGIN CERTIFICATE-----",
  "insertPolicy": "did:web:registry.example.com:policy-as-code:blocklist%40sha256%3Aaaaaaaaa"
}
$ reg layer --skip-ping --auth-url "${REGISTRY}" "${REGISTRY}/${PROJECT}/${IMAGE}@${DIGEST}" | tar -z --extract --to-stdout --wildcards --no-anchored 'policy.json' | tee insertPolicy.json
$ nodemon -e .cose --exec 'find workspace/storage/operations -name \*.cose -exec nohup sh -xc "echo {} && policy-as-code-engine insertPolicy.json {}" \;'
```

As seen in the example above the `nodemon` command in the documentation would be replaced by the `policy-as-code-engine` watching the workspace directory directly or it would continue to run the engine. The engine implements the evaluation of the JSON and executes implementations analogous or the same as the existing python files from the current revision of the docs. Effectively upleveling the existing implementation in the docs into a policy-as-code defined by the JSON workflow syntax with an associated schema and interpretation methodology of the schema.

In the above example `insertPolicy.json` would be the SCITT instance wide insert policy, which as the SCITT docs suggest could be made available at a discoverable location (webfinger?). The instance wide policy might contain per-issuer policy (or other differentiation criteria). On failure to insert (https://github.com/ietf-wg-scitt/draft-ietf-scitt-architecture/issues/62#issuecomment-1526492830) the hope would be to reply via the `error` object to the claim insert request with the JSON representation of the policy which resulted in the failure or denial. The error object could contain the specific part of the instance wide policy which caused a denial or failure for the given claim insertion.

pdxjohnny committed:
docs: registration_policies: Leverage jsonschema for evaluation of claim

Signed-off-by: John Andersen <johnandersenpdx@gmail.com>

pdxjohnny committed:
docs: registration_policies: Decode COSE messages payload before JSON load

Signed-off-by: John Andersen <johnandersenpdx@gmail.com>

OR13 reviewed (dismissed):
Looks good to me, we may need a more active maintainer on this repo to keep merging things, my python is probably not good enough for that role.

pdxjohnny committed:
Add jsonschema to dev-requirements.txt

Signed-off-by: John Andersen <johnandersenpdx@gmail.com>

pdxjohnny committed:
docs: registration_policies: Add policy_engine.sh wrapper and source policy from service parameters insertPolicy

Signed-off-by: John Andersen <johnandersenpdx@gmail.com>

johnandersen777 commented:
Thanks Orie!

I took a stab at evolving the demo towards JSON as policy over here: https://github.com/pdxjohnny/scitt-api-emulator/pull/1/files

pdxjohnny committed:
Add jsonschema to envrionment.yml

Signed-off-by: John Andersen <johnandersenpdx@gmail.com>

pdxjohnny committed:
docs: registration_polices: Convert example from blocklist to allowlist per @OR13 review

Related: https://github.com/pdxjohnny/scitt-api-emulator/pull/1#pullrequestreview-1506375585
Signed-off-by: John Andersen <johnandersenpdx@gmail.com>

OR13 commented:
Conflicts exist now, I will merge 1 week after they are resolved unless there are changes requested.

johnandersen777 commented:
Fantastic, thank you! Will resolve today

On Thu, Aug 24, 2023 at 07:22 Orie Steele ***@***.***> wrote:

> Conflicts exist now, I will merge 1 week after they are resolved unless
> there are changes requested.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/scitt-community/scitt-api-emulator/pull/27#issuecomment-1691782034>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABNMXYKJT6XME6UFHTPHXS3XW5PRXANCNFSM6AAAAAAWOESPOU>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

SteveLasker reviewed (dismissed):
LGTM

pdxjohnny committed:
docs: registration_policies: Fix policy_engine.sh wrarper to remove errantly encode jsonschema validator output into detail string

Signed-off-by: John Andersen <johnandersenpdx@gmail.com>

johnandersen777 commented:
Thanks all! I rebased in main. Looks like theres an error in the conda tests will fix asap

SteveLasker reviewed (dismissed):
LGTM

pdxjohnny committed:
tests: docs: Support for conda test execution via addition of PYTHONPATH

To ensure scitt_emulator module can be found

Signed-off-by: John Andersen <johnandersenpdx@gmail.com>

johnandersen777 commented:
The conda CI tests should be fixed now, for some reason it couldn't find the `scitt_emulator` module when executing the subprocess. Solved by setting `PYTHONPATH` to ensure to module is importable
